File: internal/slackagent/slack_context.go
package slackagent

import (
	"context"
	"fmt"
	"log/slog"
	"strings"
	"time"

	"github.com/AFK-surf/oneesama/internal/persistence"
	appconfig "github.com/AFK-surf/oneesama/pkg/config"
)

const slackWorkspaceContextsCollection = "slack_workspace_contexts"

type SlackContextRecord struct {
	ID              string                 `json:"id"`
	WorkspaceID     string                 `json:"workspaceId"`
	TeamDomain      string                 `json:"teamDomain,omitempty"`
	ChannelID       string                 `json:"channelId"`
	ChannelName     string                 `json:"channelName,omitempty"`
	ThreadTS        string                 `json:"threadTs"`
	LastUserID      string                 `json:"lastUserId,omitempty"`
	LastUserName    string                 `json:"lastUserName,omitempty"`
	LastAction      string                 `json:"lastAction,omitempty"`
	LastCommandText string                 `json:"lastCommandText,omitempty"`
	LastSessionID   string                 `json:"lastSessionId,omitempty"`
	LastMeetURL     string                 `json:"lastMeetUrl,omitempty"`
	LastTask        string                 `json:"lastTask,omitempty"`
	CommandCount    int                    `json:"commandCount"`
	RecentCommands  []RecentSlackCommand   `json:"recentCommands,omitempty"`
	ChannelBrain    string                 `json:"channelBrain,omitempty"`
	ThreadLedger    SlackThreadLedger      `json:"threadLedger,omitempty"`
	Source          SlackContextSource     `json:"source"`
	RawPublic       map[string]interface{} `json:"rawPublic,omitempty"`
	UpdatedAt       string                 `json:"updatedAt"`
	CreatedAt       string                 `json:"createdAt"`
}

type RecentSlackCommand struct {
	TS        string `json:"ts"`
	Action    string `json:"action"`
	Text      string `json:"text"`
	UserID    string `json:"userId,omitempty"`
	SessionID string `json:"sessionId,omitempty"`
	MeetURL   string `json:"meetUrl,omitempty"`
	Task      string `json:"task,omitempty"`
}

type SlackThreadLedger struct {
	LatestSessionID string `json:"latestSessionId,omitempty"`
	LatestMeetURL   string `json:"latestMeetUrl,omitempty"`
	LatestTask      string `json:"latestTask,omitempty"`
	UpdatedAt       string `json:"updatedAt,omitempty"`
}

type SlackContextSource struct {
	Kind string `json:"kind"`
	Note string `json:"note"`
}

type slackContextStore struct {
	logger     *slog.Logger
	collection *persistence.TypedCollection[SlackContextRecord]
}

func newSlackContextStore(cfg appconfig.PersistenceConfig, logger *slog.Logger) *slackContextStore {
	collection, err := persistence.OpenTyped[SlackContextRecord](persistence.Options{
		Provider:   persistence.NormalizeProvider(cfg.Provider),
		Collection: slackWorkspaceContextsCollection,
		DataDir:    cfg.DataDir,
		SQLitePath: cfg.SQLitePath,
	})
	if err != nil {
		logger.Warn("slack context store init failed", "error", err)
		return nil
	}
	return &slackContextStore{logger: logger, collection: collection}
}

func (s *slackContextStore) Remember(ctx context.Context, input AvatarCommandInput, parsed parsedAvatarCommand, session *meetingAgentSession) (*SlackContextRecord, error) {
	if s == nil || s.collection == nil {
		return nil, nil
	}
	id := slackContextID(input)
	previous, ok, err := s.collection.Get(ctx, id)
	if err != nil {
		return nil, fmt.Errorf("load slack context: %w", err)
	}
	now := time.Now().UTC().Format(time.RFC3339Nano)
	recent := RecentSlackCommand{
		TS:        now,
		Action:    firstNonEmpty(parsed.Action, "unknown"),
		Text:      strings.TrimSpace(input.Text),
		UserID:    input.UserID,
		SessionID: firstNonEmpty(sessionIDFromMeetingSession(session), parsed.SessionID),
		MeetURL:   firstNonEmpty(meetURLFromMeetingSession(session), parsed.MeetURL),
		Task:      parsed.Task,
	}
	record := SlackContextRecord{
		ID:              id,
		WorkspaceID:     firstNonEmpty(input.TeamID, "workspace"),
		TeamDomain:      input.TeamDomain,
		ChannelID:       firstNonEmpty(input.ChannelID, "channel"),
		ChannelName:     input.ChannelName,
		ThreadTS:        firstNonEmpty(input.ThreadTS, "channel-root"),
		LastUserID:      input.UserID,
		LastUserName:    input.UserName,
		LastAction:      recent.Action,
		LastCommandText: truncateSlackContextText(recent.Text, 4000),
		LastSessionID:   recent.SessionID,
		LastMeetURL:     recent.MeetURL,
		LastTask:        recent.Task,
		CommandCount:    1,
		RecentCommands:  []RecentSlackCommand{recent},
		ThreadLedger: SlackThreadLedger{
			LatestSessionID: recent.SessionID,
			LatestMeetURL:   recent.MeetURL,
			LatestTask:      recent.Task,
			UpdatedAt:       now,
		},
		Source: SlackContextSource{
			Kind: "slack-command",
			Note: "Derived from Slack mention/DM/command payload; private response_url/trigger_id/token fields are intentionally omitted.",
		},
		RawPublic: publicSlackCommandBody(input),
		UpdatedAt: now,
		CreatedAt: now,
	}
	if ok {
		record.CommandCount = previous.CommandCount + 1
		record.RecentCommands = append(append([]RecentSlackCommand(nil), previous.RecentCommands...), recent)
		if len(record.RecentCommands) > 12 {
			record.RecentCommands = record.RecentCommands[len(record.RecentCommands)-12:]
		}
		record.ChannelBrain = previous.ChannelBrain
		record.CreatedAt = firstNonEmpty(previous.CreatedAt, now)
		if record.ThreadLedger.LatestSessionID == "" {
			record.ThreadLedger.LatestSessionID = previous.ThreadLedger.LatestSessionID
		}
		if record.ThreadLedger.LatestMeetURL == "" {
			record.ThreadLedger.LatestMeetURL = previous.ThreadLedger.LatestMeetURL
		}
		if record.ThreadLedger.LatestTask == "" {
			record.ThreadLedger.LatestTask = previous.ThreadLedger.LatestTask
		}
	}
	if err := s.collection.Set(ctx, id, record); err != nil {
		return nil, fmt.Errorf("save slack context: %w", err)
	}
	return &record, nil
}

func slackContextID(input AvatarCommandInput) string {
	return strings.Join([]string{
		firstNonEmpty(input.TeamID, "workspace"),
		firstNonEmpty(input.ChannelID, "channel"),
		firstNonEmpty(input.ThreadTS, "channel-root"),
	}, ":")
}

func publicSlackCommandBody(input AvatarCommandInput) map[string]interface{} {
	return map[string]interface{}{
		"team_id":      input.TeamID,
		"team_domain":  input.TeamDomain,
		"channel_id":   input.ChannelID,
		"channel_name": input.ChannelName,
		"thread_ts":    input.ThreadTS,
		"reaction_ts":  input.ReactionTS,
		"user_id":      input.UserID,
		"user_name":    input.UserName,
		"command":      input.Command,
		"text":         input.Text,
	}
}

func sessionIDFromMeetingSession(session *meetingAgentSession) string {
	if session == nil {
		return ""
	}
	return session.ID
}

func meetURLFromMeetingSession(session *meetingAgentSession) string {
	if session == nil {
		return ""
	}
	return session.MeetingURL
}

File: packages/core/src/slack/slack-context.ts
import { createStateCollection } from "../persistence/state-provider.js";

const PRIVATE_SLACK_FIELDS = new Set(["token", "response_url", "trigger_id", "api_app_id"]);

export interface SlackContextBody {
  team_id?: string;
  team_domain?: string;
  channel_id?: string;
  channel_name?: string;
  thread_ts?: string;
  message_ts?: string;
  user_id?: string;
  user?: string;
  user_name?: string;
  command?: string;
  text?: string;
  raw?: string;
  [key: string]: unknown;
}

export interface SlackContextParsed {
  action?: string;
  sessionId?: string;
  meetUrl?: string;
  task?: string;
  flags?: Record<string, string | boolean | undefined>;
  positionals?: string[];
  validMeetUrl?: boolean;
  avatar?: string;
  botName?: string;
  dryRunJoiner?: boolean;
  startJoiner?: boolean;
  requestedMode?: string;
  allowCodeChanges?: boolean;
  [key: string]: unknown;
}

export interface SlackContextSession {
  id?: string;
  meetUrl?: string;
  [key: string]: unknown;
}

export interface SlackContextRecord {
  id: string;
  workspaceId: string;
  teamDomain: string;
  channelId: string;
  channelName: string;
  threadTs: string;
  lastUserId: string;
  lastUserName: string;
  lastAction: string;
  lastCommandText: string;
  lastSessionId: string;
  lastMeetUrl: string;
  lastTask: string;
  commandCount: number;
  recentCommands: RecentSlackCommand[];
  channelBrain: string;
  threadLedger: SlackThreadLedger;
  source: { kind: string; note: string };
  rawPublic: Record<string, unknown>;
  updatedAt: string;
  createdAt: string;
}

export interface RecentSlackCommand {
  ts: string;
  action: string;
  text: string;
  userId: string;
  sessionId: string;
  meetUrl: string;
  task: string;
}

export interface SlackThreadLedger {
  latestSessionId?: string;
  latestMeetUrl?: string;
  latestTask?: string;
  updatedAt?: string;
  [key: string]: unknown;
}

function text(value: unknown): string {
  return String(value || "").trim();
}

function commandText(body: SlackContextBody): string {
  return text(body.text || body.raw || "").slice(0, 4000);
}

function publicSlackBody(body: SlackContextBody = {}): Record<string, unknown> {
  const safe: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(body)) {
    if (PRIVATE_SLACK_FIELDS.has(key)) continue;
    safe[key] = value;
  }
  return safe;
}

interface ContextIdInput {
  workspaceId?: string;
  channelId?: string;
  threadTs?: string;
}

function contextId(input: ContextIdInput = {}): string {
  return [
    text(input.workspaceId) || "workspace",
    text(input.channelId) || "channel",
    text(input.threadTs) || "channel-root",
  ].join(":");
}

export interface CreateSlackContextProviderOptions {
  provider?: string;
  filePath?: string;
  sqlitePath?: string;
  collection?: string;
}

export interface RememberSlackCommandArgs {
  body?: SlackContextBody | Record<string, unknown>;
  parsed?: SlackContextParsed | Record<string, unknown>;
  session?: SlackContextSession | null;
}

export interface BuildSlackAgentContextArgs extends RememberSlackCommandArgs {
  remembered?: SlackContextRecord | null;
}

export function createSlackContextProvider(options: CreateSlackContextProviderOptions = {}) {
  const contexts = createStateCollection({
    provider: options.provider || (options.filePath ? "json-file" : "memory"),
    filePath: options.filePath,
    sqlitePath: options.sqlitePath,
    collection: options.collection || "slack_workspace_contexts",
  });

  function rememberCommand({
    body = {},
    parsed = {},
    session = null,
  }: RememberSlackCommandArgs = {}): SlackContextRecord {
    const b = body as SlackContextBody;
    const p = parsed as SlackContextParsed;
    const workspaceId = text(b.team_id);
    const channelId = text(b.channel_id);
    const threadTs = text(b.thread_ts || b.message_ts);
    const id = contextId({ workspaceId, channelId, threadTs });
    const previous =
      (contexts.get(id) as SlackContextRecord | undefined) || ({} as Partial<SlackContextRecord>);
    const now = new Date().toISOString();
    const recentCommand: RecentSlackCommand = {
      ts: now,
      action: p.action || "unknown",
      text: commandText(body),
      userId: text(b.user_id || b.user),
      sessionId: session?.id || p.sessionId || "",
      meetUrl: session?.meetUrl || p.meetUrl || "",
      task: p.task || "",
    };
    const recentCommands = [...(previous.recentCommands || []), recentCommand].slice(-12);
    const next: SlackContextRecord = {
      id,
      workspaceId,
      teamDomain: text(b.team_domain),
      channelId,
      channelName: text(b.channel_name),
      threadTs,
      lastUserId: recentCommand.userId,
      lastUserName: text(b.user_name),
      lastAction: recentCommand.action,
      lastCommandText: recentCommand.text,
      lastSessionId: recentCommand.sessionId,
      lastMeetUrl: recentCommand.meetUrl,
      lastTask: recentCommand.task,
      commandCount: Number(previous.commandCount || 0) + 1,
      recentCommands,
      channelBrain: previous.channelBrain || "",
      threadLedger: {
        ...previous.threadLedger,
        latestSessionId: recentCommand.sessionId || previous.threadLedger?.latestSessionId || "",
        latestMeetUrl: recentCommand.meetUrl || previous.threadLedger?.latestMeetUrl || "",
        latestTask: recentCommand.task || previous.threadLedger?.latestTask || "",
        updatedAt: now,
      },
      source: {
        kind: "slack-command",
        note: "Derived from Slack slash-command payload; private response_url/trigger_id/token fields are intentionally omitted.",
      },
      rawPublic: publicSlackBody(body),
      updatedAt: now,
      createdAt: previous.createdAt || now,
    };
    contexts.set(id, next);
    return next;
  }

  function buildAgentContext({
    body = {},
    parsed = {},
    session = null,
    remembered = null,
  }: BuildSlackAgentContextArgs = {}) {
    const b = body as SlackContextBody;
    const p = parsed as SlackContextParsed;
    const workspace: SlackContextRecord =
      remembered ||
      (contexts.get(
        contextId({
          workspaceId: b.team_id,
          channelId: b.channel_id,
          threadTs: b.thread_ts || b.message_ts,
        }),
      ) as SlackContextRecord | undefined) ||
      rememberCommand({ body, parsed, session });
    return {
      sessionId: session?.id || p.sessionId || "",
      meetUrl: session?.meetUrl || p.meetUrl || "",
      requestedBy: text(b.user_id || b.user),
      source: "slack-agent",
      slack: {
        workspaceId: workspace.workspaceId,
        teamDomain: workspace.teamDomain,
        channelId: workspace.channelId,
        channelName: workspace.channelName,
        threadTs: workspace.threadTs,
        userId: text(b.user_id || b.user),
        userName: text(b.user_name),
        command: text(b.command || "/avatar"),
        action: p.action || workspace.lastAction,
      },
      workspaceContext: {
        commandCount: workspace.commandCount,
        recentCommands: workspace.recentCommands || [],
        channelBrain: workspace.channelBrain,
        threadLedger: workspace.threadLedger,
        provenance:
          "Legacy Slack Agent D inspired: channel/thread workspace context is collected here, while private workspace memory stays behind adapters.",
      },
    };
  }

  return {
    rememberCommand,
    buildAgentContext,
    get: (id: string) => contexts.get(id) as SlackContextRecord | undefined,
    list: () =>
      (contexts.list() as SlackContextRecord[]).sort((a, b) =>
        a.updatedAt.localeCompare(b.updatedAt),
      ),
    provider: contexts.provider,
    path: contexts.path,
    collection: contexts.collection,
    close() {
      contexts.close?.();
    },
  };
}
